examples/zero_copy/c++/zero_copy_subscriber.cxx

```diff
@@ -0,0 +1,497 @@
+
+    virtual void on_sample_rejected(
+        DataReader* /*reader*/,
+        const SampleRejectedStatus& /*status*/) {}
+
+    virtual void on_liveliness_changed(
+        DataReader* /*reader*/,
+        const LivelinessChangedStatus& /*status*/) {}
+
+    virtual void on_sample_lost(
+        DataReader* /*reader*/,
+        const SampleLostStatus& /*status*/) {}
+
+    virtual void on_data_available(DataReader* reader);
+
+    ZeroCopyListener() :
+        total_latency(DDS_DURATION_ZERO),
+        received_frames(0),
+        shutdown_application(false)
+    {
+        clock = RTIHighResolutionClock_new();
+    }
+    ~ZeroCopyListener()
+    {
+        RTIHighResolutionClock_delete(clock);
+    }
+    bool start_shutdown()
+    {
+        return shutdown_application;
+    }
+  private:
+    // Sum of frame latency
+    DDS_Duration_t total_latency;
+    // Total number of received frames
+    int received_frames;
+    // Used for getting system time
+    RTIClock *clock;
+    struct RTINtpTime previous_print_time;
+    bool shutdown_application;
+
+    // Prints average latency in micro seconds
+    void print_average_latency()
+    {
+        std::cout << "Average latency: "
+                << ((total_latency.sec * 1000000)
+                        + (total_latency.nanosec / 1000)) / received_frames
+                << " microseconds" << std::endl;
+    }
+};
+
+void ZeroCopyListener::on_data_available(DataReader* reader)
+{
+    ZeroCopyDataReader *ZeroCopy_reader = NULL;
+    ZeroCopySeq data_seq;
+    SampleInfoSeq info_seq;
+    ReturnCode_t retcode;
+    int i;
+    unsigned int received_checksum = 0, computed_checksum = 0;
+    struct RTINtpTime current_time = RTI_NTP_TIME_ZERO;
+    struct RTINtpTime latency_time = RTI_NTP_TIME_ZERO;
+    struct RTINtpTime source_time = RTI_NTP_TIME_ZERO;
+
+    struct DDS_Duration_t latency;
+    ZeroCopy_reader = ZeroCopyDataReader::narrow(reader);
+    if (ZeroCopy_reader == NULL) {
+        std::cout << "DataReader narrow error" << std::endl;
+        return;
+    }
+
+    retcode = ZeroCopy_reader->take(
+        data_seq, info_seq, LENGTH_UNLIMITED,
+        ANY_SAMPLE_STATE, ANY_VIEW_STATE, ANY_INSTANCE_STATE);
+
+    if (retcode == RETCODE_NO_DATA) {
+        return;
+    } else if (retcode != RETCODE_OK) {
+        std::cout << "take error "<< retcode << std::endl;
+        return;
+    }
+
+    for (i = 0; i < data_seq.length(); ++i) {
+        if (info_seq[i].valid_data) {
+
+            // Get the frame_set_view associated with the virtual guid
+            // This is done because there may be multiple Publishers
+            // publishing frames. Each publisher creates its own
+            // shared memory segement to store frames and is identified by
+            // info_seq[i].publication_virtual_guid
+            FrameSetView *frame_set_view = get_frame_set_view(
+                    info_seq[i].publication_virtual_guid);
+
+            if (frame_set_view != NULL) {
```

Comment:
checking for the opposite condition and calling continue makes the code cleaner.